feat: adding stageMetadata validator for usdGeom

beersandrew opened the pull request:
### Description of Change(s)
- Part 2 of #3122, including upAxis & metersPerUnit checks in a new usdGeom validator structure
- added unit tests
- added associated changes to plugInfo.json & CMakeLists.txt

### Fixes Issue(s)
- #3122 

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [X] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

beersandrew committed:
feat: adding stageMetadata validator for usdGeom

- adding all validation files for usdGeom to check for valid upAxis & metersPerUnit

beersandrew committed:
doc: update header

- update header

jesschimein commented:
Filed as internal issue #USD-9787

jesschimein commented:
/AzurePipelines run

tallytalwar commented on pxr/usd/usdGeom/plugInfo.json:
```diff
@@ -390,6 +390,14 @@
+                },
+                "Validators": {
+                    "StageMetadataChecker": {
+                        "doc": "All stages should declare their 'upAxis' and 'metersPerUnit' and stages meant for consumer-level packaging should always have upAxis set to 'Y'."
```
The secondary bits of "consumer-level packaging should always have upAxis set to 'Y', wasnt that a arkit rule requirement from ComplianceChecker. I would remove this statement. (https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdUtils/complianceChecker.py#L213-L217)

It looks like you code is also not doing this check, which is correct!

beersandrew replied:
good catch, updated

tallytalwar commented on pxr/usd/usdGeom/testenv/testUsdGeomValidators.cpp:
```diff
@@ -0,0 +1,62 @@
+
+    // Verify both metersPerUnit and upAxis errors are present
+    TF_AXIOM(errors.size() == 2);
+    TF_AXIOM(errors[0].GetType() == UsdValidationErrorType::Error);
```
Other than GetMessage, you can probably loop to size of errors. And then for GetMessage you can explicitly check the 0th and 1st index?

beersandrew replied:
added these messages to their own list as well so everything can be done in the loop. let me know what you think

tallytalwar commented on pxr/usd/usdGeom/testenv/testUsdGeomValidators.cpp:
```diff
@@ -0,0 +1,62 @@
+    errors = validator->Validate(usdStage);
+
+    // Verify the errors are fixed
+    TF_AXIOM(errors.size() == 0);
```
Nice!

tallytalwar commented on pxr/usd/usdGeom/validators.cpp:
```diff
@@ -0,0 +1,45 @@
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+    static
```
Extra indentation here.

beersandrew replied:
updated

tallytalwar commented on pxr/usd/usdGeom/validators.cpp:
```diff
@@ -0,0 +1,45 @@
+                UsdGeomTokens->metersPerUnit)) {
+            errors.emplace_back(UsdValidationErrorType::Error,
+                                UsdValidationErrorSites{UsdValidationErrorSite(usdStage, SdfPath("/"))},
+                                "Stage does not specify its linear scale in "
```
Might make sense to provide the rootLayer of the Stage in the error string.

beersandrew replied:
updated

tallytalwar reviewed (commented):
Thanks @beersandrew 

Left some notes. Also might make sense to also add a test which is similar to what I have in usdShade: https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdShade/testenv/testUsdShadeValidators.cpp#L22-L46

Will sync this in as soon as these notes are addressed. Thanks.

beersandrew committed:
fix: address PR comments

- fixed description in plugInfo.json to not include ARKit rule that the upAxis must be Y
- fixed indent in validator
- refactored test to  loop through error messages
- added root layer to error messages

jesschimein commented:
/AzurePipelines run

tallytalwar reviewed (approved):
Syncing this in.

tallytalwar commented on pxr/usd/usdGeom/testenv/testUsdGeomValidators.cpp:
```diff
@@ -0,0 +1,66 @@
+            TfStringPrintf("Stage with root layer <%s> does not specify an upAxis.", rootLayerIdentifier)
+    };
+
+    for(int i = 0; i < errors.size(); ++i)
```
size() returns a size_t (unsigned_int) and the index here is an int. (I corrected this in the internal sync already, just fyi).

tallytalwar commented:
synced in, will be available in the next dev push after the release. Thanks a lot @beersandrew for this PR and other validator contributions.
